[kimxw] iP

kimxw opened the pull request:
# RapGod

> "You don't get another chance, life is no Nintendo game."  
> — Eminem

RapGod helps you manage both life and work seamlessly, ensuring you never miss a beat. It's:

- **Text-based**: Easy to use with a simple interface.
- **Easy to learn**: Get started quickly without a steep learning curve.
- **FAST**: Extremely efficient for managing tasks.

Getting started is simple:

1. Run the application.
2. Add your tasks.
3. Let RapGod handle the rest for you! 😉

### Features

- **Echo Bot**: A bot that repeats your input. Perfect for when you need a break and want to de-stress.
- **List Bot**: Manages and displays your tasks with various commands.

#### List Bot Special Commands

- **'LIST'**: Show the full list of tasks.
- **'FIND'**: Filters tasks with abc.
- **'MARK n'**: Marks the nth task as complete.
- **'UNMARK n'**: Marks the nth task as incomplete.
- **'DELETE n'**: Deletes the nth task.
- **'/BY z'**: Specifies a deadline `z`.
- **'/FROM x /TO y'**: Specifies the start and end of an event from `x` to `y`.

**Time formats accepted**:
- `dd/MM/yyyy`
- `dd/MM/yyyy HHmm`

### Example Task List

Here's an example of what your task list might look like after using the commands:
```
Displaying ListBot:

1. [T] [X] Complete project report
2. [D] [ ] Submit tax forms (by: Sep 30 2024 05:00pm)
3. [E] [X] Team meeting (from: Jul 23 2024 10:00am to: Jul 23 2024 12:00pm)
4. [T] [ ] Buy groceries
5. [D] [ ] Finish reading book (by: Aug 15 2024)
6. [E] [ ] Conference (from: Nov 01 2024 to: Nov 03 2024)
7. [T] [ ] Schedule dentist appointment
```
### To Do

- [x] JavaDoc
- [x] Update **List Bot** with new commands eg 'find'
- [ ] Add unit tests for new features

Any feeback is welcome :)

Looking forward to your review!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

kimxw committed:
Rename class

kimxw committed:
Change logo

kimxw committed:
Add greeting and exit

kimxw committed:
Adding echo support

kimxw committed:
New class for echobot

kimxw committed:
Adding list bot class

kimxw committed:
Bug fixes for list bot

kimxw committed:
List bot integration with main bot

kimxw committed:
Create Task Class

kimxw committed:
List class completion and integration with main bot

kimxw committed:
List class typo fixed

kimxw committed:
Creating 3 new types of tasks: Dealdine, Event and ToDo

kimxw committed:
Basic task differentiation functionality (without dates)

kimxw committed:
Deadline implementation for task types

kimxw committed:
Creating input and expected files and modifying runtest to use the right java class to compile.

kimxw committed:
Simple bugfix: adding extra line before each input.

kimxw committed:
Created 3 exception classes (checked exceptions) - RudeInputException and NoInputException inherit from RapGodException

kimxw committed:
Bug fixing for exception classes

kimxw committed:
Incorporating exceptions inside existing code base

kimxw committed:
Delete task

kimxw committed:
Update tests to test new functionality

kimxw committed:
AI.md

kimxw committed:
Adding enums for list bot, and javadoc for the functions

kimxw committed:
Revert "Adding enums for list bot, and javadoc for the functions"

This reverts commit cd86b4b59cf4c453ec8089b5419105b5e6f570ee.

kimxw committed:
Update EXPECTED.txt

kimxw committed:
Add enums for list bot

kimxw committed:
Create file rapgod.txt in data directory to save tasks locally

kimxw committed:
Create DataManager class for retrieval and update

kimxw committed:
Add fromString() method for tasks

kimxw committed:
Implement local data fetching with corrupted data handling; create new file if none exists

kimxw committed:
Merge branch 'branch-Level-7'

kimxw committed:
Rename classes for clarity

kimxw committed:
Delete unused classes

kimxw committed:
Modify date fields to be of type LocalDateTime

kimxw committed:
Add parsing methods in Task class

kimxw committed:
Make codebase consistent

kimxw committed:
Merge branch 'branch-Level-8'

kimxw committed:
Fix small typos

kimxw committed:
create dedicated Parser class

kimxw committed:
Encapsulate deparser inside Parser class

kimxw committed:
create TaskList for mutation of List<Task> list (field in DataManager)

kimxw committed:
Ignore data file rapgod.txt

kimxw committed:
Merge branch 'A-MoreOOP'

kimxw committed:
create packages

kimxw committed:
Merge branch 'A-Packages'

kimxw committed:
Merge remote-tracking branch 'origin/add-gradle-support'

kimxw committed:
Create test directories and files

isaactodo commented on src/main/java/rapgod/RapGod.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.Arrays;
+import java.util.Scanner;
+
+public class RapGod {
```
Nice chatbot name! 💯

isaactodo commented on src/main/java/rapgod/RapGod.java:
```diff
@@ -0,0 +1,90 @@
+        System.out.println(logo);
+        System.out.print(introduction);
+
+        while(true) {
```
No space between while(true).
```suggestion
        while (true) {
```

isaactodo commented on src/main/java/rapgod/RapGod.java:
```diff
@@ -0,0 +1,90 @@
+                break;
+            } else {
+                System.out.println("Option does not exist. Try again.");
+            }
```
Perhaps use switch statements?

isaactodo commented on src/main/java/rapgod/RapGod.java:
```diff
@@ -0,0 +1,90 @@
+            }
+
+        }
+    }
```
Perhaps add a blank line between methods? Similar issue is present in other files.

isaactodo commented on src/main/java/rapgod/RapGod.java:
```diff
@@ -0,0 +1,90 @@
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Scanner;
```
Incorrect import ordering. Similar issue present in other files.

isaactodo commented on src/main/java/rapgod/exceptions/NoInputException.java:
```diff
@@ -0,0 +1,12 @@
+package rapgod.exceptions;
```
Maybe add a blank line between the package statement and the class? Similar issue is present in other files.

isaactodo commented on src/main/java/rapgod/storage/DataManager.java:
```diff
@@ -0,0 +1,100 @@
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Pattern;
```
Incorrect import ordering.

isaactodo commented on src/main/java/rapgod/storage/DataManager.java:
```diff
@@ -0,0 +1,100 @@
+     * If the file does not exist, it creates an empty file.
+     * Validates each line against the expected format and adds valid tasks to the list.
+     *
+     * @return A list of {@link Task} objects read from the file.
```
I like the use of link tags!

isaactodo commented on src/main/java/rapgod/tasks/Deadline.java:
```diff
@@ -0,0 +1,23 @@
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task{
+    protected LocalDateTime due;
```
Use nouns for variables. 'due' sounds like an adjective.
```suggestion
    protected LocalDateTime dueDate;
```

isaactodo commented on src/main/java/rapgod/tasks/Event.java:
```diff
@@ -0,0 +1,31 @@
+import java.time.format.DateTimeFormatter;
+
+public class Event extends Task{
+    protected LocalDateTime from;
+    protected LocalDateTime to;
```
Variable names are too short, I suggest making them more descriptive.

zienxu replied:
you could rename them to startDateTime and endDateTime

isaactodo commented on src/main/java/rapgod/tasks/Deadline.java:
```diff
@@ -0,0 +1,23 @@
+    public Deadline(String description, String due) {
+        super(description);
+        this.due = Parser.parseToDateTime(due);
+    }
```
Why is there no blank line between methods? This issue is present in a lot of files.

isaactodo commented on src/main/java/rapgod/ui/ListBot.java:
```diff
@@ -0,0 +1,173 @@
+                CommandType command = CommandType.getCommand(input);
+
+                switch (command) {
+                    case LIST:
```
Incorrect indentation level for cases.

isaactodo commented on src/main/java/rapgod/ui/EchoBot.java:
```diff
@@ -0,0 +1,54 @@
+import rapgod.exceptions.NoInputException;
+
+import java.util.Scanner;
+public class EchoBot {
```
I like how you separated out echobot and listbot

isaactodo commented on src/main/java/rapgod/ui/ListBot.java:
```diff
@@ -0,0 +1,173 @@
+     */
+    public static void run() {
+
+        String initialise = """
```
Consider using nouns for variable names. 'initialise' is not a noun.

zienxu replied:
as mentioned before, you may use initialMessage

isaactodo commented on src/main/java/rapgod/ui/ListBot.java:
```diff
@@ -0,0 +1,173 @@
+                    case EVENT:
+                        String eventDesc = input.substring(0, input.toLowerCase().indexOf("/from"));
+                        String from = input.substring(input.toLowerCase().indexOf("/from") + 6, input.toLowerCase().indexOf("/to") - 1);
+                        String to = input.substring(input.toLowerCase().indexOf("/to") + 4);
```
Variable names can be more descriptive, e.g. fromDate, toDate

isaactodo commented on src/main/java/rapgod/ui/ListBot.java:
```diff
@@ -0,0 +1,173 @@
+
+                    case DEADLINE:
+                        String deadlineDesc = input.substring(0, input.toLowerCase().indexOf("/by"));
+                        String due = input.substring(input.toLowerCase().indexOf("/by") + 4);
```
Variable name due is a bit too short. I feel that dueDate would be better.

isaactodo commented on src/main/java/rapgod/utils/Parser.java:
```diff
@@ -0,0 +1,92 @@
+    public static LocalDateTime parseToDateTime(String input) {
+        DateTimeFormatter dateTimeFormatter1 = DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm");
+        DateTimeFormatter dateTimeFormatter2 = DateTimeFormatter.ofPattern("MMM dd yyyy");
+        DateTimeFormatter dateTimeFormatter3 = DateTimeFormatter.ofPattern("MMM dd yyyy h:mma");
+        DateTimeFormatter dateTimeFormatter4 = DateTimeFormatter.ofPattern("MMM dd yyyy hh:mma");
+
+        try {
+            return LocalDateTime.parse(input, dateTimeFormatter1);
+        } catch (DateTimeParseException e1) {
+            try {
+                return LocalDateTime.parse(input + " 0000", dateTimeFormatter1);
+            } catch (DateTimeParseException e2) {
+                try {
+                    LocalDate date = LocalDate.parse(input, dateTimeFormatter2);
+                    return date.atStartOfDay();
+                } catch (DateTimeParseException e3) {
+                    try {
+                        return LocalDateTime.parse(input, dateTimeFormatter3);
+                    } catch (DateTimeParseException e4) {
+                        try {
+                            return LocalDateTime.parse(input, dateTimeFormatter4);
+                        } catch (DateTimeParseException e5) {
+                            throw new IllegalArgumentException("Invalid date or time format. Please use formats like 'dd/MM/yyyy HHmm' or 'MMM dd yyyy'.");
+                        }
+                    }
+                }
+            }
+        }
+    }
```
There has got to be a better way of formatting this.

zienxu replied:
Since each block tries different date formats, you can make them into separate helper methods eg parseWithFormat1(String input)

isaactodo commented on src/main/java/rapgod/utils/Parser.java:
```diff
@@ -0,0 +1,92 @@
+     * @throws IllegalArgumentException If the input format is invalid.
+     */
+    public static LocalDateTime parseToDateTime(String input) {
+        DateTimeFormatter dateTimeFormatter1 = DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm");
+        DateTimeFormatter dateTimeFormatter2 = DateTimeFormatter.ofPattern("MMM dd yyyy");
+        DateTimeFormatter dateTimeFormatter3 = DateTimeFormatter.ofPattern("MMM dd yyyy h:mma");
+        DateTimeFormatter dateTimeFormatter4 = DateTimeFormatter.ofPattern("MMM dd yyyy hh:mma");
```
Maybe don't use numbers to distinguish names, I suggest making the names more descriptive like dateTimeFormatterSlashed

kimxw replied:
Thanks for all the comments Isaac! Will rectify the issues

isaactodo reviewed (commented):
Quite a number of coding standard violations especially in regard to spacing. Other than that, LGTM!

kimxw committed:
Add Junit test for Parser.java

kimxw committed:
Add Junit test for TaskList class

kimxw committed:
Add JUnit tests for Task class

kimxw committed:
Add empty test files for future JUnit tests

kimxw committed:
Merge branch 'A-JUnit'

kimxw committed:
Follow conding conventions

kimxw committed:
Follow coding conventions for the remaining classes

kimxw committed:
Add javadoc for Parser class

kimxw committed:
Add javadoc for ListBot and EchoBot

kimxw committed:
Add javadoc for all Task types

kimxw committed:
Add javadoc for classess in 'storage' package

kimxw committed:
Add javadoc for RapGodExceptions

kimxw committed:
Find a task by searching for a keyword in the task description

kimxw committed:
Merge branch 'branch-A-JavaDoc'

kimxw committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/rapgod/exceptions/NoInputException.java
#	src/main/java/rapgod/exceptions/RapGodException.java
#	src/main/java/rapgod/exceptions/RudeInputException.java
#	src/main/java/rapgod/storage/DataManager.java

kimxw committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/rapgod/storage/TaskList.java
#	src/main/java/rapgod/ui/ListBot.java

kimxw committed:
Fix typo

kimxw committed:
Add varargs for filterAndShowList(String ... keywords) so that users can filter by more than one word when seperated by comma

kimxw committed:
Merge branch 'branch-A-Varargs'

kimxw committed:
mark start of javaFX project

zienxu commented on src/main/java/rapgod/utils/Parser.java:
```diff
@@ -0,0 +1,102 @@
+        String info = taskString; // The description starts after the brackets
+
+        if (taskString.charAt(4) == 'D') {
+            String deadlineDesc = info.substring(11, info.toLowerCase().indexOf(" (by:"));
+            String due = info.substring(info.toLowerCase().indexOf(" (by:") + 6, info.length() - 1);
+            Deadline deadline = new Deadline(deadlineDesc, due);
+            deadline.setIsDone(isDone);
+            return deadline;
+
+        } else if (taskString.charAt(4) == 'E') {
+            String eventDesc = info.substring(11, info.toLowerCase().indexOf(" (from:"));
+            String from = info.substring(info.toLowerCase().indexOf(" (from:") + 8, info.toLowerCase().indexOf(" to:"));
+            String to = info.substring(info.toLowerCase().indexOf("to:") + 4, info.length() - 1);
+            Event event = new Event(eventDesc, from, to);
+            event.setIsDone(isDone);
+            return event;
+
+        } else {
+            ToDo todo = new ToDo(info.substring(11));
+            todo.setIsDone(isDone);
+            return todo;
+        }
```
you may consider using String.spit() to parse parts of the string. This can reduce potential indexing errors :)

zienxu commented on src/main/java/rapgod/ui/EchoBot.java:
```diff
@@ -0,0 +1,74 @@
+     */
+    public static void run() {
+
+        String initialise = """
```
can consider renaming this var to initialMessage

zienxu reviewed (commented):
LGTM. Just a few nits to fix. Really good design of the chatbot, do keep it up! very small errors with variable namings.

kimxw committed:
Update build.gradle to include JavaFX dependency

kimxw committed:
Add GUI components, controllers and style sheets

kimxw committed:
Modify existing code to give output through GUI

kimxw committed:
Close applications 2 seconds after 'bye' input

kimxw committed:
Merge branch 'branch-Level-10'

kimxw committed:
Parser class: add logging to prevent empty catch blocks

The previous implementation of the parseToDateTime method used multiple
catch blocks to handle DateTimeParseException without any logging or
handling inside the blocks. This resulted in empty catch blocks, which
can make debugging difficult and lead to silently ignored errors.

Empty catch blocks are generally bad practice as they obscure potential
errors and make debugging harder.

Add logging to each catch block to provide feedback about which parsing
format failed. This ensures that failed attempts to parse the input
string are traceable and provides better visibility into parsing issues.

Logging is used in each block because it allows us to capture and
display the exact exception message while still maintaining the fallback
parsing attempts.

This approach improves code quality without changing the structure of
the existing error-handling logic.

kimxw committed:
Refactor readMemory: Extract helper method to load task lines

The original readMemory method handles both file validation and loading
task lines, resulting in a single method that is doing too much. The
logic for reading and validating task data is mixed with the logic for
file creation, leading to a lack of separation of concerns.

This change extracts the task-loading logic into a helper method
loadLinesToList, making the readMemory method more focused and easier
to maintain. The helper method now handles validating and adding valid
tasks to the list, while ensuring that necessary directories and files
are created if they don’t exist.

This refactoring improves readability and follows the
single-responsibility principle, as the task loading is now decoupled
from file validation and creation. Additionally, it makes future changes
 to file handling or task parsing easier to implement.

No functionality is changed, but the code is cleaner and more modular,
enhancing maintainability.

kimxw committed:
Add assert statements to check assumptions

kimxw committed:
Remove redundancy by declaring class attribute

kimxw committed:
Find alternative to heavy indentation

kimxw committed:
Refactor TaskList class to use Java Streams

The TaskList class uses traditional loops to iterate over the list of
tasks for display and filtering.

Using Java Streams simplifies the code, improves readability, and aligns
with modern Java practices. It also allows for a more functional
programming approach.

Refactor the taskString() and filteredTask() methods to use Java Streams
for iteration and filtering of tasks.

This is done to reduce manual iteration and enhance maintainability.
Streams provide a more declarative approach, making the code cleaner
and easier to extend. Additionally, stream-based processing makes it
easier to parallelize tasks if needed in the future.

No changes are made to the core logic, ensuring existing functionality
is retained.

kimxw committed:
Merge pull request #3 from kimxw/branch-A-Assertions

Add assert statements to check assumptions

kimxw committed:
Merge branch 'master' into branch-A-Streams

kimxw committed:
Merge branch 'master' into branch-A-CodeQuality

kimxw committed:
Merge pull request #1 from kimxw/branch-A-CodeQuality

Branch-A-CodeQuality

kimxw committed:
Merge pull request #2 from kimxw/branch-A-Streams

Refactor TaskList class to use Java Streams

kimxw committed:
Add setters for Tasks to allow deadline modification

kimxw committed:
Add methods to allow snooze

kimxw committed:
Merge branch 'branch-B-Snooze'

kimxw committed:
Update README.md

kimxw committed:
Update build.gradle

kimxw committed:
Improve GUI

kimxw committed:
Remove console out statements

kimxw committed:
Change font in CSS style sheet

kimxw committed:
Merge branch 'branch-A-BetterGui'

kimxw committed:
Add personality

kimxw committed:
Add new exception for event end time > start time

kimxw committed:
Change command requirements to include Add if a task is to be addded

kimxw committed:
Add JUnit tests for todo

kimxw committed:
Update checkstyle

kimxw committed:
Set up product website - readme and screenshot

kimxw committed:
Change un recognised characters to ascii

kimxw committed:
Merge branch 'branch-FinalChanges'

kimxw committed:
Follow coding standard stylecheck

kimxw committed:
Add JUnit test to assert that exceptions are thrown

kimxw committed:
Update README.md

kimxw committed:
Merge branch 'master' of https://github.com/kimxw/ip

kimxw committed:
Fix bugs for release v2

kimxw committed:
Merge branch 'A-ReleaseV2'

kimxw committed:
Reduce getResponse(String input) method legnth

kimxw committed:
Delete Ui.png

kimxw committed:
Add Ui.png

kimxw committed:
Merge branch 'A-ReleaseV2'

kimxw committed:
Update README.md
